### PR TITLE
[SPARK-28042][K8S] Support using volume mount as local storage

### PR DESCRIPTION
**Pull request by chenjunjiedada**
## What changes were proposed in this pull request?

This pr is used to support using hostpath/PV volume mounts as local storage. In KubernetesExecutorBuilder.scala, the LocalDrisFeatureStep is built before MountVolumesFeatureStep which means we cannot use any volumes mount later. This pr adjust the order of feature building steps which moves localDirsFeature at last so that we can check if directories in SPARK_LOCAL_DIRS are set to volumes mounted such as hostPath, PV, or others.

## How was this patch tested?
Unit tests

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @chenjunjiedada !

**Comment by dongjoon-hyun** [comment]
ok to test

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala [hunk 1]
```diff
@@ -116,4 +115,28 @@ class LocalDirsFeatureStepSuite extends SparkFunSuite {
         .withValue(defaultLocalDir)
         .build())
   }
+
+  test("local dir on mounted volume") {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala]: This seems to fail on Jenkins. Could you check this once more? Or, do you need a change on test environment?
```
org.apache.spark.deploy.k8s.features.LocalDirsFeatureStepSuite.local dir on mounted volume
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala [hunk 2]
```diff
@@ -43,12 +43,12 @@ private[spark] class KubernetesDriverBuilder {
...
       new MountSecretsFeatureStep(conf),
       new EnvSecretsFeatureStep(conf),
-      new LocalDirsFeatureStep(conf),
       new MountVolumesFeatureStep(conf),
       new DriverCommandFeatureStep(conf),
       new HadoopConfDriverFeatureStep(conf),
       new KerberosConfDriverFeatureStep(conf),
-      new PodTemplateConfigMapStep(conf))
+      new PodTemplateConfigMapStep(conf),
+      new LocalDirsFeatureStep(conf))
```
Comment (felixcheung) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: why not move this right after `new MountVolumesFeatureStep(conf)`?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 3]
```diff
@@ -43,20 +44,34 @@ private[spark] class LocalDirsFeatureStep(
...
       .zip(resolvedLocalDirs)
       .map { case (localDirVolume, localDirPath) =>
-        new VolumeMountBuilder()
-          .withName(localDirVolume.getName)
-          .withMountPath(localDirPath)
-          .build()
+        hasVolumeMount(pod, localDirVolume.getName, localDirPath) match {
+          case true =>
+            pod.container.getVolumeMounts().asScala
+            .find(m => m.getName.equals(localDirVolume.getName)
```
Comment (felixcheung) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: please see other lines for indentation

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 4]
```diff
@@ -43,20 +44,34 @@ private[spark] class LocalDirsFeatureStep(
...
       .zipWithIndex
       .map { case (localDir, index) =>
-        new VolumeBuilder()
-          .withName(s"spark-local-dir-${index + 1}")
-          .withNewEmptyDir()
-            .withMedium(if (useLocalDirTmpFs) "Memory" else null)
-          .endEmptyDir()
-          .build()
+        val name = s"spark-local-dir-${index + 1}"
+        hasVolume(pod, name) match {
```
Comment (felixcheung) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: not a super big deal, but `hasVolume` is basically the same logic `pod.pod.getSpec().getVolumes().asScala.exists` as ` pod.pod.getSpec.getVolumes().asScala.find`

I think you can refactor this to avoid scanning multiple times.

**Review by felixcheung — COMMENTED** [review]
could we add some documentation?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala (hunk 2 above)
Comment (chenjunjiedada) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: I think some volume mounts maybe setup in PodTemplateConfigMapStep, so I move to last to ensure that.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala (hunk 3 above)
Comment (chenjunjiedada) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: ok, will check and run scalafmt.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala (hunk 1 above)
Comment (chenjunjiedada) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala]: Updated code and unit test, thanks.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala (hunk 4 above)
Comment (chenjunjiedada) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Updated in the latest commit.

**Comment by chenjunjiedada** [comment]
Hi @dongjoon-hyun and @felixcheung,

I updated the code in the latest commits, could you please take a look? Thanks.

**Comment by felixcheung** [comment]
@mccheah

**Review by liyinan926 — APPROVED** [review]
LGTM

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 5]
```diff
@@ -43,33 +44,52 @@ private[spark] class LocalDirsFeatureStep(
...
       .map { case (localDir, index) =>
-        new VolumeBuilder()
-          .withName(s"spark-local-dir-${index + 1}")
-          .withNewEmptyDir()
-            .withMedium(if (useLocalDirTmpFs) "Memory" else null)
-          .endEmptyDir()
-          .build()
+        val name = s"spark-local-dir-${index + 1}"
+        findVolume(pod, name) match {
+          case Some(volume) => volume
```
Comment (erikerlandson) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: integration tests should exercise both cases (found the volume or created it)

Comment (chenjunjiedada) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: I agree with you, there is an existing unit test for creating a local directory via emptyDir, this PR adds a unit test for the case of found volume. Both cases should already be covered. Is that ok?

**Comment by mccheah** [comment]
So my impression was that hostPath volumes are unsafe because they break the boundary of isolation between the container and the underlying machine, and the isolation between containers. I also was under the impression that containers have to be run as root to support hostPath volumes. Because of these factors I would be hesitant to advertise supporting this as a first-class feature in the Spark on Kubernetes integration. If a user so desired they can use the existing pod template feature to break glass, and this seems more like a break glass feature in my estimation.

**Comment by chenjunjiedada** [comment]
Good point @mccheah,

First, I'd like to explain that this patch is not only for hostPath volume but also it suits for other volumes such as PV, etc. it tries to adjust feature build order to adopt volumes which already be supported as first-class features. 

Second, the hostPath volume can be written by a non-root user if we change the file permission according to the description [here](https://kubernetes.io/docs/concepts/storage/volumes/#hostpath).

Lastly, I think users that care about the performance should want this since even they can define volumes inside podTemplate, the volumes are not utilized,  I have run spark-sql-perf on spark on Kubernetes with 1T scale, with the patch, most of the shuffle bounded queries can improve a lot, especially for q17, q25, q29, when using 4 disks as local storage instead of emptyDir, the query time improve more than 10X. 

In summary, this patch is just to help users easily leverage what they have to improve performance.

**Comment by chenjunjiedada** [comment]
@vanzin , could you please help to have a look?

**Comment by francoisfernando** [comment]
@mccheah 
I have hit a roadblock due to this issue trying to set the `spark.loca.dir` to a something other than `emptyDir` volumes. For example following configuration throws an exception when executor pods are provisioned.

```
--conf spark.local.dir=/spark_tmp \
--conf spark.kubernetes.executor.volumes.hostPath.spaklocal.mount.path=/spark_tmp \
--conf spark.kubernetes.executor.volumes.hostPath.spaklocal.mount.readOnly=false \
--conf spark.kubernetes.executor.volumes.hostPath.spaklocal.options.path=/tmp \
```

**Comment by chenjunjiedada** [comment]
Hi @felixcheung @mccheah , Do you think it is ready? thanks.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by chenjunjiedada** [comment]
Ping @vanzin and @felixcheung, could you please have a look?

**Comment by vanzin** [comment]
A few things:

- you explained this in comments. But the PR title explicitly talks about host paths and this is not only for host paths. That should be fixed.
- I'm a little confused about how someone would write their configuration to trigger this.

The second probably means better documentation, which you'll need to update. IIUC you list your local dirs in the configuration, and to use an existing volume you have to name it `spark-local-dir-$x` where `$x` is the index of the local dir in that array.

With the current config-based volume mounts I don't think you can do much better. But it's kinda awkward, and I'd like to propose a different approach, which relies on pod templates.

Basically, `LocalDirsFeatureStep` could detect all volumes (mounts?) with some Spark-specific label that indicates it should be used as temp storage, and update the configuration based on that. (e.g. all volume mounts with label `spark-local-dir`; I'm kinda assuming you can add labels to anything).

If you're using the old config-based mounts, you wouldn't get this feature. But I think we should be moving to pod templates as a way to configure this stuff anyway. (And I hope we can with time remove all the non-pod-template configs to make things less confusing.)

That also would not require changing the order of the feature steps (although there is no harm in doing that).

**Comment by chenjunjiedada** [comment]
@vanzin, thanks for your input.

I agree to use pod-template for setting local temporary storage. Since the configuration way still works currently (may be deprecated in future?), so the changing ordering may still be needed for a while and it is no harm anyway.  Do you mean add more document in configuration.md?

The find-and-use logic should be used for the pod-template way as well since `LocalDirFeatureSetup` doesn't make use of any volume currently. As for labeling, I don't find a way to label the volume, could you please help to provide some links?

**Comment by rvesse** [comment]
Some history - I tried to propose a similar feature in PR #22256 but this was rejected because much of the same concerns as seen in the comments here.

The "just use pod templates argument doesn't work" because `LocalDirsFeatureStep` is always going to run and set `SPARK_LOCAL_DIRS` overwriting anything the user has attempted to do in their template.

If the desire is to tell users to use pod templates then the minimum change required is the ability to disable `LocalDirsFeatureStep`

**Comment by francoisfernando** [comment]
As @rvesse has mentioned, this prevents anyone from remapping the `spark.local.dir` which I think in the greater context of Spark is a bug in Kubernetes/Spark integration.

**Comment by vanzin** [comment]
I'm not saying "just use pod templates".

I'm saying "use pod templates to mount volume into the container, and add a `spark-local-dir` tag to the mounts" or something. Spark then just needs to check which mounts from the template have that tag, and adjust the local dir configuration accordingly.

The current approach in the PR, which uses the old-style mount configs, is awkward to use, and I'd rather avoid it.

**Comment by rvesse** [comment]
@vanzin Sorry if I misconstrued your position

The idea of having a tag is a nice one, unfortunately the API doesn't have any way to do that explicitly.  I think we could do it by having a naming convention e.g. `local-dir-<name>` in lieu of that if that would be acceptable?

**Comment by vanzin** [comment]
> unfortunately the API doesn't have any way to do that explicitly

Yeah that's the part I did not get around to actually testing.

> we could do it by having a naming convention e.g. local-dir-<name>

That sounds good to me.

**Comment by vanzin** [comment]
I just checked the k8s API and it does not seem like the `VolumeMount` type has labels or anything much other than volume name / mount point.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 6]
```diff
@@ -43,33 +44,52 @@ private[spark] class LocalDirsFeatureStep(
...
+          case None =>
+            new VolumeBuilder()
+              .withName(s"spark-local-dir-${index + 1}")
```
Comment (erikerlandson) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: What is the scope of visibility on volume names "spark-local-dir-1", "spark-local-dir-2" ...
Can those names collide with volumes created by other running jobs?

**Comment by erikerlandson** [comment]
I was originally anticipating/hoping that spark's config for making use of volumes could also be managed w/out new code to detect configured volumes, e.g. with `--conf`. If there is a logistic or UX issue with that, then we should upgrade the code, but I'm trying to clarify if that is only possible approach.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala (hunk 6 above)
Comment (chenjunjiedada) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: It should be visible for global scope. Since the task distributed should run after setting up executor, so it will not collide with volumes which are built later.

**Comment by chenjunjiedada** [comment]
@rvesse @vanzin In summary, if the user specifies the directory that cannot be found in current volume (which volume name prefix is `local-dir-`), build an emptyDir volume, else use the existing volume. Is this acceptable?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala (hunk 6 above)
Comment (rvesse) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: volume names in this context are scoped to the pod spec

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala (hunk 2 above)
Comment (rvesse) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: Yep, since the intent is for users to provide custom local dir volumes via either mount volumes or pod templates it needs to appear after both

**Review by rvesse — APPROVED** [review]

**Comment by rvesse** [comment]
@erikerlandson Yes sorry I should have been clearer, I meant the K8S API

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 7]
```diff
@@ -18,7 +18,8 @@ package org.apache.spark.deploy.k8s.features
 
 import java.util.UUID
 
-import io.fabric8.kubernetes.api.model.{ContainerBuilder, HasMetadata, PodBuilder, VolumeBuilder, VolumeMountBuilder}
+import collection.JavaConverters._
+import io.fabric8.kubernetes.api.model.{ContainerBuilder, PodBuilder, Volume, VolumeBuilder, VolumeMount, VolumeMountBuilder}
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: at this point, just use the wildcard instead of a really huge line.

**Review by vanzin — COMMENTED** [review]
You still haven't really changed how awkward the configuration is. If you try to write the documentation for that, as has been requested, you'll see how difficult it is to set this up.

Instead, what you could do is look at all the mounts and see which ones match the directories in the local dirs configuration; then create empty dir volumes for the ones that don't match any mount.

But even that is still awkward.

As I suggested, what I'd like to see is a way to automatically set the local dirs based on existing mounts, without having to explicitly set the local dirs configuration.

You can do that, e.g., by looking at the existence of any mounted volume with name "spark-local-dir-*" or something; then just create the local dir configuration with that list of directories. That's even backwards compatible since for all purposes, the "spark-local-dir-" prefix was used by Spark so you would get in trouble if you named any volume that way.

If there are no mounts with that prefix, then fall back to the old behavior or creating empty dir mounts.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala (hunk 7 above)
Comment (chenjunjiedada) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: I see, Thanks.

**Comment by chenjunjiedada** [comment]
@vanzin, I was trying to summarize to ensure we are on the same page. And I think that latest commit should cover concerns.

**Comment by chenjunjiedada** [comment]
Integration test is failed, but I see it fails in other PR as well. It should be unrelated.

**Comment by vanzin** [comment]
@shaneknapp see k8s failure above (also in other PRs), seems there's something wrong with infra...

File: docs/running-on-kubernetes.md [hunk 8]
```diff
@@ -285,7 +285,16 @@ The configuration properties for mounting volumes into the executor pods use pre
 
 ## Local Storage
 
-Spark uses temporary scratch space to spill data to disk during shuffles and other operations.  When using Kubernetes as the resource manager the pods will be created with an [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume mounted for each directory listed in `SPARK_LOCAL_DIRS`.  If no directories are explicitly specified then a default directory is created and configured appropriately.
+Spark supports using volumes to spill data during shuffles and other operations. To use a volume as local storage, the volume's name should starts with `spark-local-dir-` and the mount path should be set in the spark configuration `spark.local.dir` or in the pod environment variable `SPARK_LOCAL_DIRS`, for example:
```
Comment (vanzin) [inline docs/running-on-kubernetes.md]: Why do you need both setting the volume name and the config value?

You can just build the configuration from the list of volumes that match the expected name.

If an explicit configuration is set, you should keep the previous behavior.

**Comment by shaneknapp** [comment]
> @shaneknapp see k8s failure above (also in other PRs), seems there's something wrong with infra...

i will take a look now, but in the future PLEASE open a JIRA rather than pinging me in a PR.  ;)

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
ok, i'm seeing errors when it's trying to pull the ceph nano image:

```
29e6fcdce5ba41a998b1d3d82be276ca   ceph-nano-0                             0/1       ImagePullBackOff   0          6m        172.17.0.4       minikube
```

also getting some `ErrImagePull`.  i will see if i can get more logs next time the build runs.

**Comment by shaneknapp** [comment]
test this please

**Comment by vanzin** [comment]
I think the ceph thing is #25222; but there were separate errors. I filed SPARK-28509, too.

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
test this please

File: docs/running-on-kubernetes.md (hunk 8 above)
Comment (chenjunjiedada) [reply docs/running-on-kubernetes.md]: OK, now I understood your consideration.

File: docs/running-on-kubernetes.md [hunk 9]
```diff
@@ -285,7 +285,15 @@ The configuration properties for mounting volumes into the executor pods use pre
...
+Spark supports using volumes to spill data during shuffles and other operations. To use a volume as local storage, the volume's name should starts with `spark-local-dir-`, for example:
+
+```
+--conf spark.kubernetes.driver.volumes.[VolumeType].spark-local-dir-[VolumeName].mount.path=<mount path>
+--conf spark.kubernetes.driver.volumes.[VolumeType].spark-local-dir-[VolumeName].mount.readOnly=false
+```
+
+
+If none volume is set as local storage, Spark uses temporary scratch space to spill data to disk during shuffles and other operations. When using Kubernetes as the resource manager the pods will be created with an [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir) volume mounted for each directory listed in `SPARK_LOCAL_DIRS`.  If no directories are explicitly specified then a default directory is created and configured appropriately.
```
Comment (vanzin) [inline docs/running-on-kubernetes.md]: "If no volume is set..."

Should mention the config (`spark.local.dir`) which is preferred over the env variable.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 10]
```diff
@@ -18,7 +18,8 @@ package org.apache.spark.deploy.k8s.features
 
 import java.util.UUID
 
-import io.fabric8.kubernetes.api.model.{ContainerBuilder, HasMetadata, PodBuilder, VolumeBuilder, VolumeMountBuilder}
+import collection.JavaConverters._
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: `import scala....`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala (hunk 2 above)
Comment (vanzin) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala]: Either is fine, just wanted to point out that the pod template is initialized before any step is executed (i.e. `PodTemplateConfigMapStep` is not where the template is loaded).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 11]
```diff
@@ -40,24 +41,33 @@ private[spark] class LocalDirsFeatureStep(
...
-          .withName(localDirVolume.getName)
-          .withMountPath(localDirPath)
-          .build()
-      }
+    var localDirs = findLocalDirVolumeMount(pod)
+    var localDirVolumes : Seq[Volume] = Seq()
+    var localDirVolumeMounts : Seq[VolumeMount] = Seq()
+
+    if (localDirs.isEmpty) {
+      localDirs = resolvedLocalDirs.toSeq
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Get rid of `resolvedLocalDirs` and move that statement here instead.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 12]
```diff
@@ -66,10 +76,22 @@ private[spark] class LocalDirsFeatureStep(
...
+        .withValue(localDirs.mkString(","))
         .endEnv()
       .addToVolumeMounts(localDirVolumeMounts: _*)
       .build()
     SparkPod(podWithLocalDirVolumes, containerWithLocalDirVolumeMounts)
   }
+
+  def findLocalDirVolumeMount(pod: SparkPod): Seq[String] = {
+    val localDirVolumes = pod.pod.getSpec.getVolumes.asScala
+      .filter(v => v.getName.startsWith("spark-local-dir-"))
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: nit: either:

    .filter { v => ... }

or

    .filter(_.getName()...)

Also in a few places below.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 13]
```diff
@@ -66,10 +76,22 @@ private[spark] class LocalDirsFeatureStep(
...
+
+    localDirVolumes.map { volume => pod.container.getVolumeMounts.asScala
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Move `pod.container...` to the next line.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 14]
```diff
@@ -66,10 +76,22 @@ private[spark] class LocalDirsFeatureStep(
...
+      .find(m => m.getName.equals(volume.getName)) match {
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: You can use `==` for strings in scala.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 15]
```diff
@@ -66,10 +76,22 @@ private[spark] class LocalDirsFeatureStep(
...
+          case Some(m) => m.getMountPath
+          case _ => ""
+      }
+    }.filter(s => s.length > 0)
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: You can use `localDirVolumes.flatMap` and avoid this filter.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala [hunk 16]
```diff
@@ -116,4 +115,33 @@ class LocalDirsFeatureStepSuite extends SparkFunSuite {
...
+      "",
+      false,
+      KubernetesHostPathVolumeConf("/hostPath/tmp")
+    )
+    val mountVolumeConf = KubernetesTestConf.createDriverConf(volumes = Seq(volumeConf))
+    val mountVolumeStep = new MountVolumesFeatureStep(mountVolumeConf)
+    val configuredPod = mountVolumeStep.configurePod(SparkPod.initialPod())
+
+    val sparkConf = new SparkConfWithEnv(Map())
+    val localDirConf = KubernetesTestConf.createDriverConf(sparkConf)
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala]: You can use the same `KubernetesConf` for all steps right? That's how it works when spark-submit is run.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala [hunk 17]
```diff
@@ -19,9 +19,8 @@ package org.apache.spark.deploy.k8s.features
 import io.fabric8.kubernetes.api.model.{EnvVarBuilder, VolumeBuilder, VolumeMountBuilder}
 
 import org.apache.spark.{SparkConf, SparkFunSuite}
-import org.apache.spark.deploy.k8s.{KubernetesTestConf, SparkPod}
+import org.apache.spark.deploy.k8s.{KubernetesHostPathVolumeConf, KubernetesTestConf, KubernetesVolumeSpec, SparkPod}
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala]: Use the wildcard when the import line gets too long.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala (hunk 16 above)
Comment (erikerlandson) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStepSuite.scala]: That is the idea, to keep building out one conf.  (There were some early proposals around making these stages "more functional" but those were considered less broadly familiar and less spark idiomatic)

**Comment by erikerlandson** [comment]
Just to make sure I understand the rationale - we want the option to automatically create local directories "spark-local-dir-xxx" as a simplified UX (why mess around with a second channel of config when a working dir can be created automatically from the information in the first?)

**Comment by vanzin** [comment]
> we want the option to automatically create local directories "spark-local-dir-xxx" as a simplified UX

Right. That's also more similar to other backends, where the local dirs are defined by the cluster manager themselves and users don't have to mess with the Spark configuration.

(Except here they have to, but it can be encapsulated in the pod template.)

**Comment by erikerlandson** [comment]
@vanzin thanks, I agree that makes sense

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 18]
```diff
@@ -18,7 +18,8 @@ package org.apache.spark.deploy.k8s.features
 
 import java.util.UUID
 
-import io.fabric8.kubernetes.api.model.{ContainerBuilder, HasMetadata, PodBuilder, VolumeBuilder, VolumeMountBuilder}
+import io.fabric8.kubernetes.api.model._
+import scala.collection.JavaConverters._
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Hmm, style checker should have complained about this (scala imports should be separate from others).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 19]
```diff
@@ -66,10 +76,19 @@ private[spark] class LocalDirsFeatureStep(
...
-        .withValue(resolvedLocalDirs.mkString(","))
+        .withValue(localDirs.mkString(","))
         .endEnv()
       .addToVolumeMounts(localDirVolumeMounts: _*)
       .build()
     SparkPod(podWithLocalDirVolumes, containerWithLocalDirVolumeMounts)
   }
+
+  def findLocalDirVolumeMount(pod: SparkPod): Seq[String] = {
+    val localDirVolumes = pod.pod.getSpec.getVolumes.asScala
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Hmm... do you need to list the volumes at all? Can't you just look at the mounts (since they already have the volume name)?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala [hunk 20]
```diff
@@ -66,10 +76,19 @@ private[spark] class LocalDirsFeatureStep(
...
         .withName("SPARK_LOCAL_DIRS")
-        .withValue(resolvedLocalDirs.mkString(","))
+        .withValue(localDirs.mkString(","))
         .endEnv()
       .addToVolumeMounts(localDirVolumeMounts: _*)
       .build()
     SparkPod(podWithLocalDirVolumes, containerWithLocalDirVolumeMounts)
   }
+
+  def findLocalDirVolumeMount(pod: SparkPod): Seq[String] = {
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/LocalDirsFeatureStep.scala]: Mount => Mounts

**Review by vanzin — COMMENTED** [review]
Just some small nits.

**Comment by vanzin** [comment]
Merging to master.
